tensorflow/cc/gradients/nn_grad.cc

```diff
@@ -47,6 +48,38 @@ Status SoftmaxGrad(const Scope& scope, const Operation& op,
 }
 REGISTER_GRADIENT_OP("Softmax", SoftmaxGrad);
 
+Status SoftmaxCrossEntropyWithLogitsGrad(const Scope& scope,
+                                          const Operation& op,
+                                          const std::vector<Output>&
+                                          grad_inputs,
+                                          std::vector<Output>* grad_outputs) {
+  // Softmax gradient with cross entropy logits function
+  // We multiply the backprop for cost with the gradients - op.output[1]
+  // There is no gradient for labels
+
+  auto softmaxGrad = op.output(1);
+  auto gradLoss = grad_inputs[0];
+  auto gradGrad = grad_inputs[1];
+
+  auto tempGrad = Mul(scope, gradLoss, softmaxGrad);
+
+  // TODO: Check if this sufficient, need a check for null ?
+  if (gradGrad.op().output_type(0) != 0) {
+	  auto logits = op.input(0);
+	  auto softmax = ops::Softmax(scope, logits);
+	  auto prod = ops::MatMul(scope, gradGrad, softmax);
+	  auto squeezeProd = ops::Squeeze(scope, prod);
+	  auto fProd = Sub(scope, gradGrad, squeezeProd);
+	  auto grad = Add(scope, tempGrad, fProd);
+	  grad_outputs->push_back(grad);
+	  grad_outputs->push_back(NoGradient());
```

Comment:
I am guessing `grad_outputs->push_back(NoGradient())` is for the second order gradient?

Reply:
Wait the implementation has changed in python, see:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_grad.py#L409
```python
 logits = op.inputs[0]
  if grad_grad is not None and not IsZero(grad_grad):
    softmax = nn_ops.softmax(logits)

    grad += ((grad_grad - array_ops.squeeze(
        math_ops.matmul(grad_grad[:, None, :], softmax[:, :, None]), axis=1)) *
             softmax)

  return grad, _BroadcastMul(grad_loss, -nn_ops.log_softmax(logits))
```
note that the second order gradient is now computed too

Reply:
@nietras I have modified my code accordingly but I'm not sure if I'm writing the tests correctly.  I am unable to find the Python test for the equivalent function. Can you please tell me where to find the python test. The file nn_grad_test.py doesn't have any tests.

Reply:
@pbanavara all I could find was the softmax op test:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/softmax_op_test.py

There seems to be some kind of `SoftmaxGrad` in `nn_grad.cc` to backup the python impl (hadn't seen this before):

https://github.com/tensorflow/tensorflow/blob/21c774cc1fc2aec8fd469c4f622fe7478b761e51/tensorflow/core/ops/nn_grad.cc#L25

For this I could find the following test:

https://github.com/tensorflow/tensorflow/blob/838a8f54f92452a15e3bb62a23ad5cd67e86933f/tensorflow/cc/gradients/nn_grad_test.cc#L95

Which I have no idea how actually does a proper test of this... seems pretty small to me.

Reply:
Cool. Yeah I saw these as well, but couldn't understand or make out what these tests were accomplishing. So, I'm in some sort of limbo now. I'll see how to resolve this.If I'm not able to, I'll cancel the PR. Perhaps more knowledgable people can pick up and work on this. I really feel bad for stalling this PR for more than 6 months.

Reply:
@suharshs is there not someone knowledgable that can help @pbanavara out, so we can get this done?

CC: @benoitsteiner I am going out on a limb here and nag a significant contributor, hope you or someone else can help. I apologize if not.